text/0523-model-argument-for-route-templates.md

```diff
@@ -0,0 +1,357 @@
+
+But that completely the wrong mental model! To avoid that, we put in a "Zoey
+says..." sidebar that explicitly denies that incorrect mental model, stating
+what it _is not_, without really explaining what it _is_.
+
+To fully explain what is going on, we would have to pause and take a detour to
+explain everyone's least favourite part of the Ember programming model –
+controllers. This would be extremely disruptive to the teaching flow, to say
+the least, but also serves no purpose at all.
+
+We opted to avoid going down that rabit hole and just treated it as a
+boilerplate syntax you have to type.
+
+If we are able to explain this using `@model` instead, it would make things a
+lot smoother here.
+
+We wouldn't have to go into the details of _how_ things get invoked internally.
+At the end of the day, it is still just a piece of syntax you type to get the
+job done, but at least it is consistent with the general mental model that
+`@names` means that thing was passed in, and it does not trigger the questions
+about what `this` is.
+
+It is also _not_ inconsistent with the fact that there is, in fact, a `this`
+context on the template. The usage and preference of using `@names` does not
+preclude the existance of a class (which provides the `this` context), as in
+the case of `@ember/components`.
+
+As for the concrete changes that need to be made in the documentation:
+
+* We will explain that `@model` is passed into the component from the route.
+  For example, in the `Route` class' `model` hook documentation, we can say
+  something like:
+
+  > The model hook is responsible for fetching and preparing any data that you
+  > need for your route. Ember will automatically call this hook for when
+  > entering a route, so that you can have an opportunity to do what you need to
+  > get the data you need. The object returned from this hook is known as the
+  > _model_ for the route.
+  >
+  > Note that since this is an `async` hook, if a promise is returned, it will
+  > be automatically `await`ed by Ember, and the _resolved value_ of the
+  > promise, as opposed to the promise object itself, will become the model of
+  > the route.
+  >
+  > The model object can be accessed from the route's template as using the
+  > `@model` argument. By default, the controller can also accessed the route's
+  > model from `this.model`. The latter behavior is customizable, see the
+  > _setupController_ method for details.
+  >
+  > The model of a route should be treated as read-only. For example, the
+  > controller _should not_ mutate its `this.model` property, as doing so will
+  > cause it to get out of sync with the rest of the system. Specifically,
+  > doing so will _not_ update the current URL, the `@model` argument in the
+  > template, the router service, etc. Instead, a _route transition_ should be
+  > performed.
+
+* Guides and API docs should be updated to prefer `@model` in route templates
```

Comment:
```suggestion
* Guides and API docs should be updated to prefer `@model` in route templates.
* API docs will still document the `model` property.
* We will remove any examples that uses `{{this.model}}` in templates, or replace them with `{{@model}}`. i.e. we won’t be documenting using `this.model` in templates, even thought it would “work”.
```